src/Ledger/Chain.lagda

```diff
@@ -70,42 +71,48 @@ data _⊢_⇀⦇_,NEWEPOCH⦈_ : NewEpochEnv → NewEpochState → Epoch → New
 \begin{figure*}[h]
 \begin{code}
   NEWEPOCH-New : ∀ {Γ} → let
-      open NewEpochState nes hiding (es) renaming (esFut to es) -- this rolls over esFut into es
+      open NewEpochState nes hiding (es)
+      open RatifyState fut using (es; removed) -- this rolls over the future enact state into es
       open LState ls
       -- TODO Wire CertState together with treasury and withdrawals
       open CertState certState
       open PState pState
-      removedGovActions = map GovActionDeposit (map proj₁ (fromList removed))
+
       pup = PPUpdateState.pup ppup
       deposits = UTxOState.deposits utxoSt
-      retired = retiring ⁻¹ e
-      govActionReturns' = concatMapˢ
-        (λ where
-          (gaid , gaSt) → map
-            (GovActionState.returnAddr gaSt ,_)
-            ((deposits ˢ) ⟪$⟫ ❴ GovActionDeposit gaid ❵)
-        )
-        (fromList removed)
-      govActionReturns = aggregate₊ (govActionReturns' , finiteness govActionReturns')
+
+      removedGovActions = flip concatMapˢ removed
+        (λ where (gaid , gaSt) → map
+                   (GovActionState.returnAddr gaSt ,_)
+                   ((deposits ∣ ❴ GovActionDeposit gaid ❵) ˢ))
+
+      govActionReturns =
+        aggregate₊ (map (λ where (a , _ , d) → a , d) removedGovActions , finiteness _)
+
       rewards = DState.rewards dState
       refunds   = govActionReturns ∣ dom (rewards ˢ)
       unclaimed = govActionReturns ∣ dom (rewards ˢ) ᶜ
+
+      govSt' = filter (¬? ∘ (_∈? map proj₁ removed) ∘ proj₁) govSt
+
+      retired = retiring ⁻¹ e
+
       certState' = record certState {
         pState = record pState { pools = pools ∣ retired ᶜ ; retiring = retiring ∣ retired ᶜ };
         dState = record dState { rewards = rewards ∪⁺ refunds } }
       utxoSt' = record utxoSt
         { fees = 0
-        ; deposits = deposits ∣ removedGovActions ᶜ
+        ; deposits = deposits ∣ map (proj₁ ∘ proj₂) removedGovActions ᶜ
         }
       ls' = record ls { govSt = govSt' ; utxoSt = utxoSt' ; certState = certState' }
       acnt' = record acnt { treasury = Acnt.treasury acnt + UTxOState.fees utxoSt + getCoin unclaimed }
     in
     e ≡ sucᵉ lastEpoch
     → record { currentEpoch = e ; GState gState ; NewEpochEnv Γ }
-        ⊢ ⟦ es  , [] , [] , false ⟧ʳ ⇀⦇ govSt ,RATIFY⦈ ⟦ es' , govSt' , removed , d ⟧ʳ
+        ⊢ ⟦ es , ∅ , false ⟧ʳ ⇀⦇ govSt' ,RATIFY⦈ fut'
     -- TODO: remove keys that aren't in the CC from the hot key map
     ────────────────────────────────
-    Γ ⊢ nes ⇀⦇ e ,NEWEPOCH⦈ ⟦ e , acnt' , ls' , es , es' ⟧ⁿᵉ
+    Γ ⊢ nes ⇀⦇ e ,NEWEPOCH⦈ ⟦ e , acnt' , ls' , es , fut' ⟧ⁿᵉ
```

Comment:
Currently in the implementation we store just the `EnactState` inside `NewEpochState`

Reply:
I think it makes more sense to do it like that here as well, since we only need to use the `removed` set inside `NEWEPOCH`

Reply:
same goes for `delay`

Reply:
what I meant is that in the EpochState we store an EnactState where in the spec we have a RatifyState. When the EPOCH rule is called, we create an ephemeral RatifyState inside the EPOCH rule to pass to RATIFY and then use the `removed` field to remove the governance actions that need to be removed. Then we move the new `EnactState` to `EpochState` and forget about the rest of the `RatifyState` once we're done executing the EPOCH rule.

Reply:
Yes, that's what used to be the case. The issue is that to know the value of `removed`, we have to do the entire `RATIFY` computation which we can only do once we've done the incremental stake distribution calculation. So we can't look at `removed` immediately, we have to wait until the next epoch boundary to use it. So we can either add a new field (say `futRemoved`) that stores them for one epoch, or we just store the entire result of `RATIFY` (which is what I did here).